File: src/algebra2_rational_equations_easy.py
"""
algebra2 - rational_equations (easy)
Generated: 2026-02-22T04:24:21.474848
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 4)
    
    if problem_type == 1:
        # Type 1: Simple rational equation with one fraction equal to a constant
        # Example: x/a = b, solve for x
        a_val = nonzero(-8, 8)
        b_val = nonzero(-8, 8)
        
        # Answer is x = a*b
        ans_val = a_val * b_val
        
        eq = Eq(x / a_val, b_val)
        
        return problem(
            question=f"Solve for $x$: ${latex(eq)}$",
            answer=ans_val,
            difficulty=(1000, 1100),
            topic="algebra2/rational_equations",
            solution=steps(
                f"Start with ${latex(eq)}$",
                f"Multiply both sides by ${a_val}$",
                f"$x = {b_val} \\cdot {a_val}$",
                f"$x = {ans_val}$"
            )
        )
    
    elif problem_type == 2:
        # Type 2: Rational equation with x in numerator: (ax + b)/c = d
        a_val = nonzero(-5, 5)
        b_val = nonzero(-8, 8)
        c_val = nonzero(-6, 6)
        d_val = nonzero(-6, 6)
        
        # Solve: ax + b = c*d, so x = (c*d - b)/a
        numerator = c_val * d_val - b_val
        if numerator % a_val == 0:
            ans_val = numerator // a_val
        else:
            ans_val = Rational(numerator, a_val)
        
        eq = Eq((a_val*x + b_val) / c_val, d_val)
        
        return problem(
            question=f"Solve for $x$: ${latex(eq)}$",
            answer=ans_val,
            difficulty=(1100, 1200),
            topic="algebra2/rational_equations",
            solution=steps(
                f"Start with ${latex(eq)}$",
                f"Multiply both sides by ${c_val}$",
                f"${latex(a_val*x + b_val)} = {c_val * d_val}$",
                f"${latex(a_val*x)} = {c_val * d_val - b_val}$",
                f"$x = {latex(ans_val)}$"
            )
        )
    
    elif problem_type == 3:
        # Type 3: Two fractions equal: a/x = b/c
        a_val = nonzero(-9, 9)
        b_val = nonzero(-9, 9)
        c_val = nonzero(-9, 9)
        
        # Cross multiply: a*c = b*x, so x = a*c/b
        numerator = a_val * c_val
        if numerator % b_val == 0:
            ans_val = numerator // b_val
        else:
            ans_val = Rational(numerator, b_val)
        
        eq = Eq(a_val / x, b_val / c_val)
        
        return problem(
            question=f"Solve for $x$: ${latex(eq)}$",
            answer=ans_val,
            difficulty=(1200, 1300),
            topic="algebra2/rational_equations",
            solution=steps(
                f"Start with ${latex(eq)}$",
                f"Cross multiply: ${a_val} \\cdot {c_val} = {b_val} \\cdot x$",
                f"${a_val * c_val} = {b_val}x$",
                f"$x = \\frac{{{a_val * c_val}}}{{{b_val}}}$",
                f"$x = {latex(ans_val)}$"
            )
        )
    
    else:
        # Type 4: Simple proportion with variable in denominator: a/(x+b) = c
        a_val = nonzero(-9, 9)
        b_val = nonzero(-8, 8)
        c_val = nonzero(-6, 6)
        
        # Solve: a = c(x+b), a = cx + cb, cx = a - cb, x = (a - cb)/c
        numerator = a_val - c_val * b_val
        if numerator % c_val == 0:
            ans_val = numerator // c_val
        else:
            ans_val = Rational(numerator, c_val)
        
        eq = Eq(a_val / (x + b_val), c_val)
        
        return problem(
            question=f"Solve for $x$: ${latex(eq)}$",
            answer=ans_val,
            difficulty=(1200, 1300),
            topic="algebra2/rational_equations",
            solution=steps(
                f"Start with ${latex(eq)}$",
                f"Multiply both sides by $(x + {b_val})$",
                f"${a_val} = {c_val}(x + {b_val})$",
                f"${a_val} = {c_val}x + {c_val * b_val}$",
                f"${c_val}x = {a_val - c_val * b_val}$",
                f"$x = {latex(ans_val)}$"
            )
        )

emit(generate())

File: src/algebra2_rational_equations_easy.jl
# algebra2 - rational_equations (easy)
# Generated: 2026-03-08T20:16:10.801636

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra2/rational_equations")
    
    # Choose problem type
    prob_type = choice(1:4)
    
    if prob_type == 1
        # Simple rational equation: a/(x+b) = c, solve for x
        # ELO ~700-900
        a = nonzero(-15, 15)
        b = nonzero(-12, 12)
        c = nonzero(-8, 8)
        ans = (a // c) - b
        
        # Make sure denominator doesn't vanish at solution
        while abs(ans + b) < 0.01
            b = nonzero(-12, 12)
            ans = (a // c) - b
        end
        
        lhs = a / (x + b)
        
        problem(
            question = "Solve for \\(x\\): \\[$(tex(lhs)) = $(c)\\]",
            answer = ans,
            difficulty = (750, 900),
            solution = steps(
                sol("Given equation", lhs ~ c),
                "Multiply both sides by \\($(tex(x + b))\\): \\($(a) = $(c)($(tex(x + b)))\\)",
                "Expand: \\($(a) = $(tex(expand(c*(x + b))))\\)",
                "Solve for \\(x\\): \\($(tex(c*x)) = $(a - c*b)\\)",
                sol("Answer", ans)
            ),
            time = 90
        )
        
    elseif prob_type == 2
        # Equation with two rational terms: a/(x+b) + c/(x+d) = e
        # ELO ~900-1100
        b = nonzero(-10, 10)
        d = nonzero(-10, 10)
        while d == b
            d = nonzero(-10, 10)
        end
        
        # Pick a solution that doesn't make denominators zero
        ans = randint(-15, 15)
        while ans == -b || ans == -d
            ans = randint(-15, 15)
        end
        
        # Work backwards: pick a, c, then compute e
        a = nonzero(-8, 8)
        c = nonzero(-8, 8)
        e_val = substitute(a/(x+b) + c/(x+d), x => ans)
        
        lhs = a/(x+b) + c/(x+d)
        
        problem(
            question = "Solve for \\(x\\): \\[$(tex(lhs)) = $(tex(e_val))\\]",
            answer = ans,
            difficulty = (900, 1100),
            solution = steps(
                sol("Given equation", lhs ~ e_val),
                "Multiply both sides by \\($(tex((x+b)*(x+d)))\\) (the LCD)",
                "Simplify: \\($(a)($(tex(x+d))) + $(c)($(tex(x+b))) = $(tex(e_val * (x+b)*(x+d)))\\)",
                "Expand and collect like terms",
                "Solve the resulting linear or quadratic equation",
                sol("Answer", ans)
            ),
            time = 150
        )
        
    elseif prob_type == 3
        # Rational equation with x in numerator: (ax+b)/(x+c) = d
        # ELO ~800-1000
        a = nonzero(-8, 8)
        b = randint(-15, 15)
        c = nonzero(-12, 12)
        d = nonzero(-7, 7)
        
        # Solve: ax + b = d(x + c) => ax + b = dx + dc => (a-d)x = dc - b
        if a == d
            # Pick different values to avoid infinite solutions
            a = nonzero(-8, 8)
            d = nonzero(-7, 7)
            while a == d
                d = nonzero(-7, 7)
            end
        end
        
        ans = (d*c - b) // (a - d)
        
        lhs = (a*x + b) / (x + c)
        
        problem(
            question = "Solve for \\(x\\): \\[$(tex(lhs)) = $(d)\\]",
            answer = ans,
            difficulty = (800, 1000),
            solution = steps(
                sol("Given equation", lhs ~ d),
                "Multiply both sides by \\($(tex(x + c))\\): \\($(tex(a*x + b)) = $(d)($(tex(x + c)))\\)",
                "Expand right side: \\($(tex(a*x + b)) = $(tex(expand(d*(x + c))))\\)",
                "Collect terms: \\($(tex((a-d)*x)) = $(d*c - b)\\)",
                sol("Answer", ans)
            ),
            time = 120
        )
        
    else  # prob_type == 4
        # Cross-multiplication: a/(x+b) = c/(x+d)
        # ELO ~850-1050
        b = nonzero(-10, 10)
        d = nonzero(-10, 10)
        while d == b
            d = nonzero(-10, 10)
        end
        
        a = nonzero(-12, 12)
        c = nonzero(-12, 12)
        
        # Solve: a(x+d) = c(x+b) => ax + ad = cx + cb => (a-c)x = cb - ad
        if a == c
            c = nonzero(-12, 12)
            while a == c
                c = nonzero(-12, 12)
            end
        end
        
        ans = (c*b - a*d) // (a - c)
        
        # Verify solution doesn't make denominators zero
        while abs(ans + b) < 0.01 || abs(ans + d) < 0.01
            b = nonzero(-10, 10)
            d = nonzero(-10, 10)
            while d == b
                d = nonzero(-10, 10)
            end
            ans = (c*b - a*d) // (a - c)
        end
        
        lhs = a / (x + b)
        rhs = c / (x + d)
        
        problem(
            question = "Solve for \\(x\\): \\[$(tex(lhs)) = $(tex(rhs))\\]",
            answer = ans,
            difficulty = (850, 1050),
            solution = steps(
                sol("Given equation", lhs ~ rhs),
                "Cross-multiply: \\($(a)($(tex(x+d))) = $(c)($(tex(x+b)))\\)",
                "Expand: \\($(tex(expand(a*(x+d)))) = $(tex(expand(c*(x+b))))\\)",
                "Collect like terms: \\($(tex((a-c)*x)) = $(c*b - a*d)\\)",
                sol("Answer", ans)
            ),
            time = 120
        )
    end
end